validator/client/wait_for_activation.go

```diff
@@ -66,44 +66,52 @@ func (v *validator) WaitForActivation(ctx context.Context) error {
 			Error("Stream broken while waiting for activation. Reconnecting...")
 		// Reconnection attempt backoff, up to 60s.
 		time.Sleep(time.Second * time.Duration(mathutil.Min(uint64(attempts), 60)))
-		return v.WaitForActivation(incrementRetries(ctx))
+		return v.WaitForActivation(incrementRetries(ctx), accountsChangedChan)
 	}
 	for {
-		res, err := stream.Recv()
-		// If the stream is closed, we stop the loop.
-		if errors.Is(err, io.EOF) {
-			break
-		}
-		// If context is canceled we stop the loop.
-		if ctx.Err() == context.Canceled {
-			return errors.Wrap(ctx.Err(), "context has been canceled so shutting down the loop")
-		}
-		if err != nil {
-			traceutil.AnnotateError(span, err)
-			attempts := getStreamAttempts(ctx)
-			log.WithError(err).WithField("attempts", attempts).
-				Error("Stream broken while waiting for activation. Reconnecting...")
-			// Reconnection attempt backoff, up to 60s.
-			time.Sleep(time.Second * time.Duration(mathutil.Min(uint64(attempts), 60)))
-			return v.WaitForActivation(incrementRetries(ctx))
-		}
-		valActivated := v.checkAndLogValidatorStatus(res.Statuses)
+		select {
+		case <-accountsChangedChan:
+			// Accounts (keys) changed, restart the process.
+			return v.WaitForActivation(ctx, accountsChangedChan)
+		default:
+			res, err := stream.Recv()
+			// If the stream is closed, we stop the loop.
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			// If context is canceled we return from the function.
+			if ctx.Err() == context.Canceled {
+				return errors.Wrap(ctx.Err(), "context has been canceled so shutting down the loop")
+			}
+			if err != nil {
+				traceutil.AnnotateError(span, err)
+				attempts := getStreamAttempts(ctx)
+				log.WithError(err).WithField("attempts", attempts).
```

Comment:
Debug log perhaps?

Reply:
I did not touch this code. I can change it to debug, but I guess the reason for `WithError` was that we have `err != nil`.